asllib/doc/SubprogramCalls.tex

```diff
@@ -112,150 +91,198 @@ \subsubsection{Formally}
         in $\tenv$ yields $\True$\ProseOrTypeError;
-  \item applying $\checkcalleeparams$ to $\calleeparams$ to check they have correct types
-        with respect to $\eqsthree$ in $\tenv$ yields $\True$\ProseOrTypeError;
-  \item applying $\annotateretty$ to $\calltype$ and $\callee.\funcreturntype$
+  \item define $\eqs$ as the association of declared parameter names in $\funcsig.\funcparameters$ with actual parameters $\paramsone$;
+  \item applying $\checkargstypesat$ to $\argtypes$ and $\eqs$
+        to check that the actual arguments have correct types with respect to $\funcsig.\funcargs$
+        in $\tenv$ yields $\True$\ProseOrTypeError;
+  \item applying $\annotateretty$ to $\eqs$, $\calltype$, and $\funcsig.\funcreturntype$
         to check that the two call types match and to substitute actual parameter arguments in the formal return type
         yields $\rettyopt$ \ProseOrTypeError;
-  \item define $\eqs$ as $\eqsthree$.
+  \item define $\vcall$ as the call with name $\namep$, parameters taken from $\paramsone$, arguments $\vargs$, and call type $\funcsig.\funcsubprogramtype$.
 \end{itemize}
 \subsubsection{Formally}
 \begin{mathpar}
 \inferrule{
-  \unziplist(\callerargstyped) = (\callerargtypes, \vargsone)\\
-  {
-    \begin{array}{r}
-    \subprogramforname(\tenv, \name, \callerargtypes) \typearrow \\
-  (\eqsone, \nameone, \callee) \OrTypeError
-    \end{array}
-  }\\
-  \checktrans{\callee.\subprogramtype = \calltype}{\MismatchedReturnValue} \checktransarrow \True\OrTypeError\\\\
-  \equallength(\callee.\funcargs, \vargsone) \typearrow \aritymatch\\
-  \checktrans{\aritymatch}{\CallBadArity} \checktransarrow \True \OrTypeError\\\\
+  \unziplist(\texttt{typed\_args}) = (\argtypes, \args)\\
+    \subprogramforname(\tenv, \name, \argtypes) \typearrow
+  (\nameone, \funcsig) \OrTypeError \\
   {
-  \annotateparameterdefining\left(
+  \hyperlink{def-checktrans}{\textfunc{check}}\left(
     \begin{array}{l}
-      \tenv,\\
-      \callee.\funcargs, \\
-      \callerargstyped, \\
-      \callee.\funcparameters
-    \end{array}
+      \funcsig.\funcsubprogramtype = \calltype \; \lor \\
+      (\funcsig.\funcsubprogramtype = \STGetter \\ \quad \land \; \calltype = \STFunction)
+    \end{array},
+    \MismatchedReturnValue
     \right)
-    \typearrow \eqsthreep \OrTypeError
-  }\\
-  \eqsthree \eqdef \eqsthreep \concat \eqsone\\
-  \checkargstypesat(\tenv, \callee.\funcargs, \callerargtypes, \eqsthree) \typearrow \True \OrTypeError\\\\
-  \checkcalleeparams(\tenv, \callee.\funcparameters, \eqsthree) \typearrow \True \OrTypeError\\\\
-  \annotateretty(\tenv, \calltype, \callee.\funcreturntype) \typearrow \rettyopt \OrTypeError
+  } \checktransarrow \True\OrTypeError\\\\
+  \insertstdlibparam(\funcsig, \params, \argtypes) \typearrow \paramsone\\\\
+  \equallength(\funcsig.\funcparameters, \paramsone) \typearrow \texttt{param\_arity\_match} \\\\
```

Comment:
it's better to define a macro for `param_arity_match`. Otherwise, error tend to creep in.